### DISCUSSION TITLE
Conversion to galactic coordinates?

### DISCUSSION THREAD

**Discussion by cjheath** [cosinekitty/astronomy#112] [Ideas]
With the rise in popularity of backyard radioastronomy, it would be wonderful to be able to calculate galactic coordinates. Especially the alt-az of specific extra-solar-system objects for a given place and time on Earth.

**Comment by cosinekitty** [comment]
Hi Clifford! That is a great idea. I did not know that backyard radioastronomy was a thing. Cool! 

It sounds like the specific use case is to convert from galactic coordinates of a distant radio source to horizontal coordinates for an observer at a given location on the Earth. I suppose this is for the purpose of aiming a radio dish?

Here are my initial thoughts for an approach I would be willing to implement. Astronomy Engine already supports four different orientation systems:

- EQJ: Earth equatorial at the J2000 epoch
- EQD: Earth equatorial for a given date and time
- ECL: Ecliptic
- HOR: Horizontal for a given date and a given location on the Earth

Astronomy Engine also provides functions for converting coordinates from any of these orientations back and forth to any of the others. I can add support for galactic coordinates (GAL), although I don't think I want to keep adding a complete set of conversions from every orientation to every other, because it would greatly expand the number of functions to document and translate among the supported programming languages.

Fortunately, GAL and EQJ are both invariant systems, meaning they are fixed and do not vary with time or location on the Earth. I think the best approach would be for Astronomy Engine to provide the fixed rotation matrix that converts from GAL to EQJ and back. Then I will provide some demo code that shows the following conversion sequence:

1. Given galactic coordinates, use the GAL_to_EQJ rotation matrix to calculate equatorial J2000 coordinates.
2. Use existing functions to convert EQJ to horizontal coordinates (altitude and azimuth) for a given date, time, latitude, and longitude.

And I would include another demo to do the inverse.

Such a demo would also allow people to figure out how to convert back and forth between galactic coordinates and the other systems.

What do you think?

**Reply by matheo** [reply]
> although I don't think I want to keep adding a complete set of conversions from every orientation to every other

Agreed! for particular applications this criteria fits nicely. Like
1. Convert the input to the coordinates required to make the processing
2. Run the particular processing
3. Apply any additional conversion to the result if needed

Just my thoughts after reading your proposal Don :)

**Reply by cjheath** [reply]
That sounds perfect! Most people seem to say "just use Stellarium" to get pointing directions, but that's difficult to automate. I started researching the required math, but 3D transforms make my head spin :)

Of course, it would also work with a standard database of bright stars, so the optical folk with automated telescope steering would also benefit. And anyone wanting to make charts like this could use that too: https://heavens-above.com/skychart2.aspx?lat=-33.758&lng=151.2058&loc=Unnamed&alt=114&tz=AEST

**Reply by cjheath** [reply]
See also the Amateur Radio Astronomy group on FB: https://www.facebook.com/groups/1819174114777651/

**Reply by cosinekitty** [reply]
Thank you @cjheath .  That helps. I have already started working on this in a new branch called `galactic`. I also asked to join that Facebook group, so I'm sure it will be fun and educational for me to see what people discuss there.

Do you have a sense of what programming languages amateur radio astronomers would find most helpful? I'm just curious, because I will implement the galactic coordinate support in all the languages Astronomy Engine supports: C/C++, C#, Python, and JavaScript.

**Reply by cjheath** [reply]
Personally I use C/C++, but I expect Python would be more widely used - it's pretty standard in scientific work. C# and Javascript are bonus, anyone wanting more can built it on C.

**Comment by cosinekitty** [comment]
I have updated the master branch with new functions to assist converting between galactic coordinates and all the other coordinate systems. I provided a [C demo program](https://github.com/cosinekitty/astronomy/blob/master/demo/c/galactic.c) and a [Python demo program](https://github.com/cosinekitty/astronomy/blob/master/demo/python/galactic.py) to illustrate converting from galactic coordinates to horizontal coordinates for an observer at a given time and geographic location. I verified these demos using the JPL Horizons online tool.

@cjheath : The demos do not correct for atmospheric refraction because the formulas I have assume visible light. I have no idea what refraction effect the atmosphere has on radio wavelengths, so these demos calculate without any refraction correction. Do you know if there is a model for atmospheric refraction in radio astronomy?

**Comment by cjheath** [comment]
That's  great news, Don. Many thanks.

I don't think refraction will affect radio astronomy, and if it does, it's only noticeable at the shortest wavelengths on the largest dishes, not amateur ones. Nevertheless I have posted a message to the Facebook group to see if anyone knows the answer.
